### PR TITLE
Add Scala 2.13 builds for Spark/v3.2

### PR DESCRIPTION
**Pull request by fqaiser94**
Resolves https://github.com/apache/iceberg/issues/3372

# What changes were proposed in this pull request?

- Added a build flag for specifying scalaVersion: `./gradlew -DscalaVersion=2.13 projects`
- Added new projects for scala 2.13: 
  - `:iceberg-spark:iceberg-spark-3.2_2.13`
  - `:iceberg-spark:iceberg-spark-extensions-3.2_2.13`
  - `:iceberg-spark:iceberg-spark-runtime-3.2_2.13`

# What approach was taken? 

Iceberg spark libraries will be built for Scala 2.12 and 2.13 using the same source files. 
This is simpler to maintain than having 2 sets of nearly identical source files. 
The only downside is you need to ensure your code works across both versions, which is where the majority of the relatively simple changes in this PR come from. 
Of particular note; a new dependency was introduced [scala-collection-compat](https://github.com/scala/scala-collection-compat) which makes some Scala 2.13 APIs available on 2.12.

# What other approaches (if any) were considered? 

I did look into existing Gradle plugins for cross-building scala ([gradle-crossbuild-scala](https://github.com/prokod/gradle-crossbuild-scala),  [gradle-scala-multiversion-plugin](https://github.com/ADTRAN/gradle-scala-multiversion-plugin)) but neither is particularly popular, and I had issues setting them up. 
Also, note that I'm pretty new to Gradle so feel free to suggest other approaches if you know better. 

# How was this patch tested?

Existing unit tests must pass for scala 2.12 and 2.13 build.

### CODE REVIEW

**Comment by wypoon** [comment]
@fqaiser94, you can edit your personal access token and under scopes, check the "workflow" box. Then you can update the github workflow.

**Comment by fqaiser94** [comment]
@wypoon thanks! 
Also, are you able to approve running workflows on this branch? 
I'd like to see if everything passes on CI at least :)

**Comment by wypoon** [comment]
> @wypoon thanks! Also, are you able to approve running workflows on this branch? I'd like to see if everything passes on CI at least :)

I am not a committer and have no privileges.

**Comment by kbendick** [comment]
@fqaiser94 I'm not able to turn them on either, but if you `./gradlew clean check  -DflinkVersions=1.12,1.13,1.14 -DsparkVersions=2.4,3.0,3.1,3.2 -DhiveVersions=2,3` locally on Java 8, that's more or less every test. Some of those don't support Java 11 though.

**Comment by rdblue** [comment]
Great to see someone working on this. Thank you, @fqaiser94!

File: .github/workflows/spark-ci.yml
Comment (rdblue) [inline .github/workflows/spark-ci.yml:127]: Only 2.12 is supported for Spark 3.1, so I don't think that we want to test that. Won't this fail when the 3.1_2.13 project is tested?

File: spark/v3.2/build.gradle
Comment (rdblue) [inline spark/v3.2/build.gradle:87]: Only Scala versions should be embedded directly. Everything else can still use versions.props because you can have multiple entries as long as the Scala library has the Scala version in the artifact name like this one (`scala-collection-compat_2.13`).

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:56]: Looks like this was incorrect before, but can you remove the wildcard import here?

File: spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala
Comment (rdblue) [reply spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSparkSqlExtensionsParser.scala:57]: Actually, it looks like the collection converters do use wildcard import. We must have bypassed the style check for this, so let's keep it the way it is (since there are existing places that do it).

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:186]: Using the Scala API in Java is really ugly. I think that's why we prefer to use the Java API and convert unless it is performance critical. Is this change needed for Scala 2.13 or can we continue to convert?

**Comment by rdblue** [comment]
This looks really close to being ready to commit to me. I'd like to separate the 2.13 tests out into a separate CI run so that we don't cause 3.1 and 3.2 tests to take twice as long.

File: .github/workflows/spark-ci.yml
Comment (fqaiser94) [reply .github/workflows/spark-ci.yml:153]: yes, it will absolutely fail for 3.1_2.13. 
Fixed this now, we have an entirely separate workflow for scala 2.13 builds.

File: spark/v3.2/build.gradle
Comment (fqaiser94) [reply spark/v3.2/build.gradle]: done.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (fqaiser94) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:186]: I fully agree with you that "using the Scala API in Java is really ugly”, but in this case, I fear this is the better choice. 

I thought it would be really simple, all we want to do is convert a java Map into an *immutable* Scala map. 
So, we start off by converting a java Map into a *mutable* Scala map:  
> scala.collection.mutable.Map<String, String> myMutableMap = JavaConverters.mapAsScalaMapConverter(myJavaMap).asScala();`

And now we can convert a *mutable* Scala map into an *immutable* Scala map using the handy-dandy `.toMap` method. 
But wait, the `.toMap` method takes an implicit argument `ev`: 
> def toMap[K, V](implicit ev: A <:< (K, V)): immutable.Map[K, V]

This is basically a compiler trick to make sure we’re calling `.toMap` on a collection of tuples (otherwise the method doesn’t make sense). 
Normally, in Scala, this evidence is implicitly supplied. 
Since we’re calling this method from Java, we don’t have the normal implicit mechanisms available so we have to construct and supply this evidence by hand. 
In the past we were able to supply this evidence by using `scala.Predef.conforms()` but this API was removed entirely in 2.13 (it was already deprecated in 2.12). 
AFAICT, in scala 2.13, this implicit `ev` value of type `<:<` is provided by `scala.Predef.refl()` in in Scala 2.13. 
However, this method is not available in scala 2.12. 
Right now, the only other alternative I can think of is having two separate files for this code for 2.12 and 2.13, but that seems overkill for something so small, hence I would argue this is the better choice. 

Hope that makes sense. Open to thoughts.

Comment (fqaiser94) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java]: I've added one more change [here](https://github.com/apache/iceberg/pull/4009/commits/440bc10b89dfe85893c7e8d2b150b6a491cf73bc) so at least it looks a little less gnarly :)

Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java]: Okay, so the problem is that we can't convert directly to an Immutable Scala map, and the Scala API's use of implicits makes converting mutable to immutable really difficult. That seems reasonable. I'm not sure what we were doing before... maybe we didn't need an Immutable map?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:184]: This does look much better. Thank you!

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks great to me! I'll wait for tests to pass and then merge.

**Comment by rdblue** [comment]
Looks like some odd test failures. Somehow Scala can't compile when all Spark versions are enabled? Maybe this is a reuse problem from having multiple Scala versions in the same build? I wouldn't have a problem with adding a flag for Scala versions and running the build checks and Javadoc on just 2.12 for now, as long as we have full tests for 2.13 in a separate workflow.

**Comment by fqaiser94** [comment]
>  Maybe this is a reuse problem from having multiple Scala versions in the same build?

Yes, I suspect this is exactly what it is. 
In the latest commit, I’ve added a build flag to specify the scalaVersion. 
By default, we will use scala 2.12 but you can switch over to 2.13 as needed e.g. `./gradlew -DscalaVersion=2.13 projects`. 

Incidentally, this approach is similar to what they do over in the apache/kafka repo.

File: .github/workflows/flink-ci.yml
Comment (rdblue) [inline .github/workflows/flink-ci.yml:76]: Is this change needed? If there are no Spark versions, then it should be a noop to set the Scala version to nothing.

File: .github/workflows/java-ci.yml
Comment (rdblue) [inline .github/workflows/java-ci.yml:80]: Can't this be the default Scala version so we don't need to set it here?

File: .github/workflows/publish-snapshot.yml
Comment (rdblue) [inline .github/workflows/publish-snapshot.yml:42]: Here as well, we probably don't need to change this line because 2.12 should be the default.

File: .github/workflows/flink-ci.yml
Comment (fqaiser94) [reply .github/workflows/flink-ci.yml:76]: Technically no, but I was just following the pattern in the rest of the line. 
I'll get rid of it :)

File: settings.gradle
Comment (rdblue) [inline settings.gradle:141]: It looks like we should never set `scalaVersion` to "" because it will cause build failures. Just leaving it as 2.12 unless we override seems reasonable to me.

File: .github/workflows/flink-ci.yml
Comment (rdblue) [reply .github/workflows/flink-ci.yml:76]: Yeah, I think that would avoid changing a few files.

File: .github/workflows/java-ci.yml
Comment (fqaiser94) [reply .github/workflows/java-ci.yml]: removed

File: spark/v3.2/build.gradle
Comment (rdblue) [inline spark/v3.2/build.gradle:27]: Can you fix this to use the original indentation?

File: spark/v3.2/build.gradle
Comment (rdblue) [inline spark/v3.2/build.gradle:30]: Can you revert the change that renames sparkProjects to projects? That way we have fewer conflict opportunities.

**Review by rdblue — APPROVED** [review]
Just a couple minor things, but otherwise this looks ready to go once tests are passing.

**Comment by fqaiser94** [comment]
@rdblue all comments addressed, can you give the CI a bump once more?

**Comment by fqaiser94** [comment]
huzzah, all green

**Comment by rdblue** [comment]
Thanks, @fqaiser94!

**Pull request merged**
